akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala

```diff
@@ -96,11 +101,36 @@ import akka.util.ConstantFun
 
   override def unstash(ctx: scaladsl.ActorContext[T], behavior: Behavior[T],
                        numberOfMessages: Int, wrap: T ⇒ T): Behavior[T] = {
-    val iter = new Iterator[T] {
-      override def hasNext: Boolean = StashBufferImpl.this.nonEmpty
-      override def next(): T = wrap(StashBufferImpl.this.dropHead())
-    }.take(numberOfMessages)
-    Behavior.interpretMessages[T](behavior, ctx, iter)
+    if (isEmpty)
+      behavior // optimization
+    else {
+      val iter = new Iterator[T] {
+        override def hasNext: Boolean = StashBufferImpl.this.nonEmpty
+        override def next(): T = wrap(StashBufferImpl.this.dropHead())
+      }.take(numberOfMessages)
+      interpretUnstashedMessages(behavior, ctx, iter)
+    }
+  }
+
+  private def interpretUnstashedMessages(behavior: Behavior[T], ctx: TypedActorContext[T], messages: Iterator[T]): Behavior[T] = {
+    @tailrec def interpretOne(b: Behavior[T]): Behavior[T] = {
+      val b2 = Behavior.start(b, ctx)
+      if (!Behavior.isAlive(b2) || !messages.hasNext) b2
+      else {
+        val nextB = try {
+          messages.next() match {
```

Comment:
One thing I'd proposed in my version is the ability to also provide the failed message to the error context, from `messages.next()` if it helped in a user understanding what failed and why

Reply:
So that `UnstashException` would have it. But that could also be too heavy.

Reply:
We avoid `.toString` on unknown things like messages because that could have a very bad performance impact on production systems if toString is badly implemented (could be a huge string). It could log the message class but not sure it helps that much. If we would do that we should also do it for ordinary supervision. Skipping it for now, can be added later if needed.